picard/ui/logview.py

```diff
@@ -97,31 +125,159 @@ def restoreWindowState(self, position, size):
             self.move(pos)
         self.resize(config.persist[size])
 
+    def highlight(self, searchString, cursor=None):
+        # adapted from http://doc.qt.io/qt-5/qtuitools-textfinder-example.html
+        if cursor:
+            highlightCursor = cursor
+        else:
+            highlightCursor = QtGui.QTextCursor(self.doc)
+
+        colorFormat = highlightCursor.charFormat()
+        colorFormat.setForeground(QtGui.QColor('green'))
+
+        while not highlightCursor.isNull() and not highlightCursor.atEnd():
+            highlightCursor = self.doc.find(searchString, highlightCursor)
+            if not highlightCursor.isNull():
+                highlightCursor.movePosition(QtGui.QTextCursor.WordRight,
+                                             QtGui.QTextCursor.KeepAnchor)
+                highlightCursor.mergeCharFormat(colorFormat)
+
+
 
 class LogView(LogViewCommon):
 
     options = [
         config.Option("persist", "logview_position", QtCore.QPoint()),
         config.Option("persist", "logview_size", QtCore.QSize(560, 400)),
+        config.Option("persist", "logview_verbosity", LogViewCommon.default_verbosity)
     ]
 
+    _log_level_names = OrderedDict([
+        (log.LOG_INFO,      N_('Info')),
+        (log.LOG_WARNING,   N_('Warning')),
+        (log.LOG_ERROR,     N_('Error')),
+        (log.LOG_DEBUG,     N_('Debug')),
+    ])
+
     def __init__(self, parent=None):
         title = _("Log")
         logger = log.main_logger
         LogViewCommon.__init__(self, title, logger, parent=parent)
+        self.verbosity = config.persist['logview_verbosity']
         self.restoreWindowState("logview_position", "logview_size")
+        self.hbox = QtWidgets.QHBoxLayout(self)
+        self.vbox.addLayout(self.hbox)
+
         cb = QtWidgets.QCheckBox(_('Debug mode'), self)
         cb.setChecked(QtCore.QObject.tagger._debug)
         cb.stateChanged.connect(self.toggleDebug)
-        self.vbox.addWidget(cb)
+        self.hbox.addWidget(cb)
+
+        self.verbosity_menu = QtWidgets.QMenu(self)
+        for level, label in self._log_level_names.items():
+            act = QtWidgets.QAction(_(label), self.verbosity_menu)
+            act.setCheckable(True)
+            act.setChecked(level in self.verbosity)
+            act.triggered.connect(partial(self._verbosity_changed, level))
+            self.verbosity_menu.addAction(act)
+
+        self.verbosity_menu_button = QtWidgets.QPushButton(_("Verbosity"))
```

Comment:
Any reason for having the gui layouts part here instead of a ui file?

Reply:
Nope, but since menus are generated, you'll not have much more than few widgets in ui file anyway.